api/s2n.h

```diff
@@ -421,6 +421,97 @@ extern int s2n_connection_set_client_auth_type(struct s2n_connection *conn, s2n_
 
+/**
+ * Sets the initial number of session tickets to send after a handshake.
+ *
+ * @param config A pointer to the config object.
+ * @param num The number of session tickets that will be sent.
+ */
+S2N_API
+extern int s2n_config_set_initial_ticket_count(struct s2n_config *config, uint8_t num);
+
+/**
+ * Increases the amount of session tickets to send after a handshake.
+ *
+ * @param conn A pointer to the s2n_connection object.
+ * @param num The number of additional session tickets to send.
+ */
+S2N_API
+extern int s2n_connection_add_new_tickets_to_send(struct s2n_connection *conn, uint8_t num);
+
+/**
+ * Sets the keying material lifetime for session tickets so that one session doesn't get re-used ad infinitum.
+ * The default value is one week.
+ *
+ * @param conn A pointer to the s2n_connection object.
+ * @param lifetime_in_secs Lifetime of keying material in seconds.
+ */
+S2N_API
+extern int s2n_connection_set_server_keying_material_lifetime(struct s2n_connection *conn, uint32_t lifetime_in_secs);
+
+struct s2n_session_ticket;
+
+/**
+ * Callback function for receiving a session ticket
+ *
+ * # Safety
+ *
+ * `ctx` MUST be cast into the same type of pointer that was originally created.
+ * `ticket` is an internal pointer that can only be used until this function exits.
+ *
+ * @param conn A pointer to the s2n_connection object.
+ * @param ctx Context for the callback.
+ * @param ticket Pointer to the received session ticket object.
+ */
+typedef int (*s2n_session_ticket_fn)(struct s2n_connection *conn, void *ctx, struct s2n_session_ticket *ticket);
+
+/**
+ * Sets a session ticket callback to be called when a client receives a new session ticket.
+ *
+ * # Safety
+ *
+ * `callback` MUST cast `ctx` into the same type of pointer that was originally created.
+ * `ctx` MUST live for at least as long as it is set on the config.
+ *
+ * @param config A pointer to the config object.
+ * @param callback The function that should be called when the callback is triggered.
+ * @param ctx The context to be passed when the callback is called.
+ */
+S2N_API
+extern int s2n_config_set_session_ticket_cb(struct s2n_config *config, s2n_session_ticket_fn callback, void *ctx);
+
+/**
+ * Gets length of session ticket from a s2n_session_ticket object.
+ *
+ * @param ticket Pointer to session ticket object.
+ * @param data_len Pointer to where the length of the session ticket will be stored.
+ */
+S2N_API
+extern int s2n_session_ticket_get_data_len(struct s2n_session_ticket *ticket, size_t *data_len);
+
+/**
+ * Gets the session ticket data from a s2n_session_ticket object.
+ *
+ * # Safety
+ * `data` must have enough memory to store the session ticket data, as the session ticket will be copied
+ * into `data`, if this function executes successfully.
+ *
+ * @param ticket Pointer to session ticket object.
+ * @param max_data_len Maximum length of data that can be written to the data pointer.
```

Comment:
what happens if someone calls this with `1` or some small invalid value and the data is more than that value? what is the expected behavior then?

how would the caller know the size of the data prior to calling this?

Reply:
The function would error if max_data_len < session ticket size and no data would be copied into the data pointer. A user would call the s2n_session_ticket_get_data_len() function to know the size of the data.

 I'm waffling on whether this information needs to be in s2n.h. I did point out in the [usage guide](https://github.com/aws/s2n-tls/pull/2877/files#diff-a074eb126585a13e10851e90d69b3ced02387b0738238eddf5720d89cb77bd65R1596) that you should use the session_ticket getters to get more information about the ticket. Do you think that is not clear enough?

Reply:
thanks for the explanation. yea i think documenting it in usage should be sufficient since we expect users to go look at that doc.